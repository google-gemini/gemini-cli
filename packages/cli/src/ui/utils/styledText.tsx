/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { Text } from 'ink';

/**
 * Renders text with styled interpolations for i18n.
 *
 * This implements the "Semantic Interpolation Pattern" for React Ink environments:
 * - Provides complete semantic context for translators
 * - Avoids fragmented translation keys
 * - Supports complex styling in terminal environments
 * - Compatible with React Ink's Text component limitations
 *
 * @param text - The translated text with {key} placeholders
 * @param styleMap - Map of keys to their styled React components
 * @param defaultColor - Default text color for the container
 * @returns Styled text component or throws error on mismatch
 * @throws Error when placeholders and styleMap don't match
 *
 * @example
 * // Translation: "Shell mode: Execute commands via {symbol} (e.g., {example})"
 * renderStyledText(t('shellMode'), {
 *   symbol: <Text bold color="purple">!</Text>,
 *   example: <Text bold color="purple">!npm start</Text>
 * }, 'white')
 */
export function renderStyledText(
  text: string,
  styleMap: Record<string, React.ReactNode>,
  defaultColor?: string,
): React.ReactElement {
  // Extract all placeholders from the text
  const placeholderMatches = text.match(/\{[^}]+\}/g) || [];
  const extractedPlaceholders = placeholderMatches.map((match) =>
    match.slice(1, -1),
  );

  // Get styleMap keys
  const styleMapKeys = Object.keys(styleMap);

  // Check for mismatches and provide detailed error messages
  const missingInStyleMap = extractedPlaceholders.filter(
    (key) => !styleMapKeys.includes(key),
  );
  const unusedInStyleMap = styleMapKeys.filter(
    (key) => !extractedPlaceholders.includes(key),
  );

  if (missingInStyleMap.length > 0 || unusedInStyleMap.length > 0) {
    const errors: string[] = [];

    if (missingInStyleMap.length > 0) {
      errors.push(
        `Missing style mappings for placeholders: {${missingInStyleMap.join('}, {')}}`,
      );
    }

    if (unusedInStyleMap.length > 0) {
      errors.push(`Unused style mappings: ${unusedInStyleMap.join(', ')}`);
    }

    throw new Error(
      `renderStyledText mismatch: ${errors.join('. ')}. ` +
        `Text: "${text}". ` +
        `Expected placeholders: [${extractedPlaceholders.join(', ')}]. ` +
        `Provided style mappings: [${styleMapKeys.join(', ')}].`,
    );
  }

  const parts = text.split(/(\{[^}]+\})/);

  return (
    <Text color={defaultColor}>
      {parts
        .map((part, _index) => {
          const match = part.match(/^\{([^}]+)\}$/);
          if (match) {
            const key = match[1];
            return styleMap[key]; // Now guaranteed to exist due to validation above
          }
          return part;
        })
        .map((element, idx) => (
          <React.Fragment key={idx}>{element}</React.Fragment>
        ))}
    </Text>
  );
}
