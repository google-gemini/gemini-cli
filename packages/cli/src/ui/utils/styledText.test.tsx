/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/** @vitest-environment jsdom */

import { render } from 'ink-testing-library';
import { describe, it, expect } from 'vitest';
import { Text } from 'ink';
import { renderStyledText } from './styledText.js';

describe('I18n Styled Text Solution', () => {
  it('should render complete semantic text with styled interpolations', () => {
    const translatedText =
      'Add context: Use {symbol} to specify files for context (e.g., {example})';

    const component = renderStyledText(
      translatedText,
      {
        symbol: (
          <Text bold color="purple">
            @
          </Text>
        ),
        example: (
          <Text bold color="purple">
            @src/myFile.ts
          </Text>
        ),
      },
      'white',
    );

    const { lastFrame } = render(component);

    const output = lastFrame();

    // Should contain the complete semantic meaning
    expect(output).toContain('Add context');
    expect(output).toContain('Use');
    expect(output).toContain('specify files');
    expect(output).toContain('@');
    expect(output).toContain('@src/myFile.ts');
  });

  it('should apply correct styling attributes to interpolated components', () => {
    const translatedText = 'Use {symbol} to specify {example}';

    const component = renderStyledText(
      translatedText,
      {
        symbol: (
          <Text bold color="purple">
            @
          </Text>
        ),
        example: (
          <Text bold color="purple">
            files
          </Text>
        ),
      },
      'white',
    );

    // Verify component structure contains correct styling attributes
    const componentString = JSON.stringify(component);

    // Should contain Text components with correct props
    expect(componentString).toContain('"bold":true');
    expect(componentString).toContain('"color":"purple"');

    // Should contain the correct text content
    expect(componentString).toContain('"@"');
    expect(componentString).toContain('"files"');
  });

  it('should provide complete context for translators', () => {
    const translatedText =
      'Shell mode: Execute shell commands via {symbol} (e.g., {example}) or use natural language (e.g. {natural})';

    // Translator sees the complete sentence structure
    expect(translatedText).toContain('Shell mode: Execute shell commands');
    expect(translatedText).toContain('or use natural language');

    // Only styling placeholders, not semantic content
    expect(translatedText).toContain('{symbol}');
    expect(translatedText).toContain('{example}');
    expect(translatedText).not.toContain('!'); // Symbol not in translation
  });

  it('should throw error when style mappings are missing', () => {
    const translatedText = 'Use {symbol} and {missing} here';

    expect(() => {
      renderStyledText(translatedText, {
        symbol: <Text bold>@</Text>,
        // 'missing' key not provided
      });
    }).toThrowError(
      /renderStyledText mismatch: Missing style mappings for placeholders: \{missing\}/,
    );
  });

  it('should throw error when unused style mappings exist', () => {
    const translatedText = 'Use {symbol} here';

    expect(() => {
      renderStyledText(translatedText, {
        symbol: <Text bold>@</Text>,
        unused: <Text>Never used</Text>,
      });
    }).toThrowError(/renderStyledText mismatch: Unused style mappings: unused/);
  });

  it('should throw error with detailed information when both missing and unused mappings exist', () => {
    const translatedText = 'Use {symbol} and {missing} here';

    expect(() => {
      renderStyledText(translatedText, {
        symbol: <Text bold>@</Text>,
        unused1: <Text>Never used 1</Text>,
        unused2: <Text>Never used 2</Text>,
        // 'missing' key not provided
      });
    }).toThrowError(
      /renderStyledText mismatch: Missing style mappings for placeholders: \{missing\}\. Unused style mappings: unused1, unused2/,
    );
  });

  it('should provide complete error context with text and mappings', () => {
    const translatedText = 'Hello {name}!';

    expect(() => {
      renderStyledText(translatedText, {
        wrongKey: <Text>Wrong</Text>,
      });
    }).toThrowError(
      /Text: "Hello \{name\}!"\. Expected placeholders: \[name\]\. Provided style mappings: \[wrongKey\]/,
    );
  });

  describe('Regex Pattern Testing', () => {
    it('should correctly split text with single placeholder', () => {
      const { lastFrame } = render(
        renderStyledText('Hello {name}!', {
          name: <Text bold>World</Text>,
        }),
      );

      expect(lastFrame()).toContain('Hello');
      expect(lastFrame()).toContain('World');
      expect(lastFrame()).toContain('!');
    });

    it('should handle multiple placeholders correctly', () => {
      const { lastFrame } = render(
        renderStyledText('{greeting} {name}, welcome to {place}!', {
          greeting: <Text color="green">Hello</Text>,
          name: <Text bold>Alice</Text>,
          place: <Text color="blue">CLI</Text>,
        }),
      );

      const output = lastFrame();
      expect(output).toContain('Hello');
      expect(output).toContain('Alice');
      expect(output).toContain('CLI');
      expect(output).toContain('welcome to');
    });

    it('should handle consecutive placeholders', () => {
      const { lastFrame } = render(
        renderStyledText('{first}{second}{third}', {
          first: <Text>A</Text>,
          second: <Text>B</Text>,
          third: <Text>C</Text>,
        }),
      );

      expect(lastFrame()).toContain('ABC');
    });

    it('should require all placeholders to have style mappings', () => {
      // Now that we have error detection, all placeholders must have mappings
      const { lastFrame } = render(
        renderStyledText('Use {valid} and {provided}', {
          valid: <Text bold>OK</Text>,
          provided: <Text>Found</Text>,
        }),
      );

      const output = lastFrame();
      expect(output).toContain('OK'); // {valid} -> OK
      expect(output).toContain('Found'); // {provided} -> Found
    });

    it('should handle empty placeholders', () => {
      const { lastFrame } = render(
        renderStyledText('Test {} and {empty}', {
          empty: <Text>Filled</Text>,
        }),
      );

      const output = lastFrame();
      expect(output).toContain('Test');
      expect(output).toContain('{}'); // Empty placeholder kept as-is
      expect(output).toContain('Filled');
    });

    it('should handle placeholders with special characters', () => {
      const { lastFrame } = render(
        renderStyledText('Use {symbol-1} and {key_2} and {number3}', {
          'symbol-1': <Text>Symbol</Text>,
          key_2: <Text>Key</Text>,
          number3: <Text>Number</Text>,
        }),
      );

      const output = lastFrame();
      expect(output).toContain('Symbol');
      expect(output).toContain('Key');
      expect(output).toContain('Number');
    });

    it('should handle nested braces correctly', () => {
      const { lastFrame } = render(
        renderStyledText('Code: {code} with escaped braces', {
          code: <Text bold>function() return true</Text>,
        }),
      );

      const output = lastFrame();
      expect(output).toContain('function() return true');
      expect(output).toContain('escaped braces');
    });

    it('should handle text with no placeholders', () => {
      const { lastFrame } = render(
        renderStyledText('Plain text with no placeholders', {}),
      );

      expect(lastFrame()).toContain('Plain text with no placeholders');
    });

    it('should handle only placeholders with no surrounding text', () => {
      const { lastFrame } = render(
        renderStyledText('{only}', {
          only: <Text bold>Placeholder</Text>,
        }),
      );

      expect(lastFrame()).toContain('Placeholder');
    });

    it('should handle complex real-world example', () => {
      const complexText =
        'Shell mode: Execute {type} commands via {symbol} (e.g., {example}) or use {method} (e.g. {natural}).';

      const { lastFrame } = render(
        renderStyledText(complexText, {
          type: <Text italic>shell</Text>,
          symbol: (
            <Text bold color="purple">
              !
            </Text>
          ),
          example: (
            <Text bold color="purple">
              !npm run start
            </Text>
          ),
          method: <Text italic>natural language</Text>,
          natural: <Text italic>start server</Text>,
        }),
      );

      const output = lastFrame();
      expect(output).toContain('Shell mode: Execute');
      expect(output).toContain('shell');
      expect(output).toContain('!');
      expect(output).toContain('!npm run start');
      expect(output).toContain('natural language');
      // Account for potential line wrapping in Ink terminal output
      expect(output).toContain('start');
      expect(output).toContain('server');
    });

    it('should handle debugging - check exact split behavior', () => {
      // This test helps us understand the regex split behavior
      const text = 'Test {missing} here';
      const { lastFrame } = render(
        renderStyledText(text, {
          missing: <Text>REPLACED</Text>,
        }),
      );

      const output = lastFrame();
      expect(output).toContain('Test');
      expect(output).toContain('REPLACED');
      expect(output).toContain('here');
      expect(output).not.toContain('{missing}');
    });
  });
});
