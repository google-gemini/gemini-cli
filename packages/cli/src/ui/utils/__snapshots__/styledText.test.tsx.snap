// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`I18n Styled Text Solution > Error Handling Golden Tests > should provide detailed error structure for missing mappings (golden test) > missing-mapping-error.json 1`] = `
"{
  "message": "renderStyledText mismatch: Missing style mappings for placeholders: {missing}. Text: \\"Use {symbol} and {missing}\\". Expected placeholders: [symbol, missing]. Provided style mappings: [symbol].",
  "name": "Error",
  "cause": null
}"
`;

exports[`I18n Styled Text Solution > Error Handling Golden Tests > should provide detailed error structure for unused mappings (golden test) > unused-mapping-error.json 1`] = `
"{
  "message": "renderStyledText mismatch: Unused style mappings: unused. Text: \\"Use {symbol}\\". Expected placeholders: [symbol]. Provided style mappings: [symbol, unused].",
  "name": "Error",
  "cause": null
}"
`;

exports[`I18n Styled Text Solution > Golden Test Suite - Complete Component Serialization > should handle complex multi-placeholder with different styles (golden test) > multi-placeholder-complex.json 1`] = `
"{
  "key": null,
  "props": {
    "children": [
      {
        "key": "0",
        "props": {
          "children": "Execute "
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "1",
        "props": {
          "children": {
            "key": null,
            "props": {
              "italic": true,
              "color": "blue",
              "children": "shell"
            },
            "_owner": null,
            "_store": {}
          }
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "2",
        "props": {
          "children": " commands via "
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "3",
        "props": {
          "children": {
            "key": null,
            "props": {
              "bold": true,
              "color": "purple",
              "children": "!"
            },
            "_owner": null,
            "_store": {}
          }
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "4",
        "props": {
          "children": " (e.g., "
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "5",
        "props": {
          "children": {
            "key": null,
            "props": {
              "bold": true,
              "color": "purple",
              "children": "!npm run start"
            },
            "_owner": null,
            "_store": {}
          }
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "6",
        "props": {
          "children": ")"
        },
        "_owner": null,
        "_store": {}
      }
    ]
  },
  "_owner": null,
  "_store": {}
}"
`;

exports[`I18n Styled Text Solution > Golden Test Suite - Complete Component Serialization > should handle nested styling and special characters (golden test) > nested-styling.json 1`] = `
"{
  "key": null,
  "props": {
    "children": [
      {
        "key": "0",
        "props": {
          "children": "Code: "
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "1",
        "props": {
          "children": {
            "key": null,
            "props": {
              "bold": true,
              "underline": true,
              "children": "function() return true"
            },
            "_owner": null,
            "_store": {}
          }
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "2",
        "props": {
          "children": " with "
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "3",
        "props": {
          "children": {
            "key": null,
            "props": {
              "bold": true,
              "color": "yellow",
              "children": "important"
            },
            "_owner": null,
            "_store": {}
          }
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "4",
        "props": {
          "children": ""
        },
        "_owner": null,
        "_store": {}
      }
    ]
  },
  "_owner": null,
  "_store": {}
}"
`;

exports[`I18n Styled Text Solution > Golden Test Suite - Complete Component Serialization > should handle only placeholders with no surrounding text (golden test) > only-placeholder.json 1`] = `
"{
  "key": null,
  "props": {
    "children": [
      {
        "key": "0",
        "props": {
          "children": ""
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "1",
        "props": {
          "children": {
            "key": null,
            "props": {
              "bold": true,
              "color": "red",
              "children": "Placeholder"
            },
            "_owner": null,
            "_store": {}
          }
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "2",
        "props": {
          "children": ""
        },
        "_owner": null,
        "_store": {}
      }
    ]
  },
  "_owner": null,
  "_store": {}
}"
`;

exports[`I18n Styled Text Solution > Golden Test Suite - Complete Component Serialization > should handle simple single placeholder (golden test) > simple-placeholder.json 1`] = `
"{
  "key": null,
  "props": {
    "children": [
      {
        "key": "0",
        "props": {
          "children": "Hello "
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "1",
        "props": {
          "children": {
            "key": null,
            "props": {
              "bold": true,
              "color": "green",
              "children": "World"
            },
            "_owner": null,
            "_store": {}
          }
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "2",
        "props": {
          "children": "!"
        },
        "_owner": null,
        "_store": {}
      }
    ]
  },
  "_owner": null,
  "_store": {}
}"
`;

exports[`I18n Styled Text Solution > Golden Test Suite - Complete Component Serialization > should handle text with no placeholders (golden test) > no-placeholders.json 1`] = `
"{
  "key": null,
  "props": {
    "color": "cyan",
    "children": [
      {
        "key": "0",
        "props": {
          "children": "Plain text with no placeholders"
        },
        "_owner": null,
        "_store": {}
      }
    ]
  },
  "_owner": null,
  "_store": {}
}"
`;

exports[`I18n Styled Text Solution > should apply correct styling attributes to interpolated components (golden test) > styled-text-component.json 1`] = `
"{
  "key": null,
  "props": {
    "color": "white",
    "children": [
      {
        "key": "0",
        "props": {
          "children": "Use "
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "1",
        "props": {
          "children": {
            "key": null,
            "props": {
              "bold": true,
              "color": "purple",
              "children": "@"
            },
            "_owner": null,
            "_store": {}
          }
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "2",
        "props": {
          "children": " to specify "
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "3",
        "props": {
          "children": {
            "key": null,
            "props": {
              "bold": true,
              "color": "purple",
              "children": "files"
            },
            "_owner": null,
            "_store": {}
          }
        },
        "_owner": null,
        "_store": {}
      },
      {
        "key": "4",
        "props": {
          "children": ""
        },
        "_owner": null,
        "_store": {}
      }
    ]
  },
  "_owner": null,
  "_store": {}
}"
`;
